lib/s3routes/routesUtils.js

```diff
@@ -49,71 +49,145 @@ function okHeaderResponse(headers, response, httpCode, log) {
+        response.writeHead(errCode.code,
+                           { 'Content-type': 'application/xml' });
+        return response.end(xmlStr, 'utf8', () => {
+            log.end().info('responded with error XML', {
+                httpCode: response.statusCode,
+            });
+        });
+    },
+};
+
+const JSONResponseBackend = {
+
+    /**
+     * okJSONResponse - Response with JSON body
+     * @param {string} json - JSON body as string
+     * @param {object} response - http response object
+     * @param {object} log - Werelogs logger
+     * @param {object} additionalHeaders -- additional headers to add
+     *   to response
+     * @return {object} response - response object with additional headers
      */
-    const xml = [];
-    xml.push(
-        '<?xml version="1.0" encoding="UTF-8"?>',
-        '<Error>',
-        `<Code>${errCode.message}</Code>`,
-        `<Message>${errCode.description}</Message>`,
-        '<Resource></Resource>',
-        `<RequestId>${log.getSerializedUids()}</RequestId>`,
-        '</Error>'
-    );
-    const xmlStr = xml.join('');
-    const bytesSent = Buffer.byteLength(xmlStr);
-    log.addDefaultFields({
-        bytesSent,
-    });
-    if (corsHeaders) {
-        // eslint-disable-next-line no-param-reassign
-        corsHeaders['Content-Type'] = 'application/xml';
-        // eslint-disable-next-line no-param-reassign
-        corsHeaders['Content-Length'] = xmlStr.length;
-    }
-    setCommonResponseHeaders(corsHeaders, response, log);
-    response.writeHead(errCode.code, { 'Content-type': 'application/xml' });
-    return response.end(xmlStr, 'utf8', () => {
-        log.end().info('responded with error XML', {
-            httpCode: response.statusCode,
+    okResponse: function okJSONResponse(json, response, log,
+                                        additionalHeaders) {
+        const bytesSent = Buffer.byteLength(json);
+        log.trace('sending success json response');
+        log.addDefaultFields({
+            bytesSent,
         });
-    });
-}
+        setCommonResponseHeaders(additionalHeaders, response, log);
+        response.writeHead(200, { 'Content-type': 'application/json' });
+        log.debug('response http code', { httpCode: 200 });
+        log.trace('json response', { json });
+        return response.end(json, 'utf8', () => {
+            log.end().info('responded with JSON', {
+                httpCode: response.statusCode,
+            });
+        });
+    },
+
+    errorResponse: function errorJSONResponse(errCode, response, log,
+                                              corsHeaders) {
+        log.trace('sending error json response', { errCode });
+        /*
+         {
+             "code": "NoSuchKey",
+             "message": "The resource you requested does not exist",
+             "resource": "/mybucket/myfoto.jpg",
+             "requestId": "4442587FB7D0A2F9"
+         }
+         */
+        const jsonStr =
```

Comment:
We need to stringify because AWS client expects the `code` and `message` fields, they are not a direct mapping of arsenal error properties.

I agree for the request ID, I basically copied what is output in the XML version, but if you think it's not needed (which I think too) I can remove. I was thinking it was in the XML because some clients expect the request id to be there. (for JSON since it's for internal needs it does not apply).